examples/async/s3_cp.ml

```diff
@@ -0,0 +1,390 @@
+        "AWS4-HMAC-SHA256 Credential=%s,SignedHeaders=%s,Signature=%s"
+        creds signed_headers (Compat.cstruct_to_hex_string signature)
+    in
+    [("Authorization", auth_header);]
+
+end
+
+module S3 = struct
+
+  type conf = {
+    region : region;
+    aws_access_key : string;
+    aws_secret_key : string;
+  } with sexp
+
+  let make_request ?body conf ~meth ~bucket ~objekt =
+    (* )let open Cohttp.Request in*)
+    let host_str = region_host_string conf.region in
+    let uri = Printf.sprintf "https://%s/%s/%s" host_str bucket objekt
+              |> Uri.of_string in
+    let time = Time.now () in
+    (* If PUT add content length *)
+    let headers = match meth with
+      | `PUT -> begin
+          let length = Option.value_map ~f:(String.length) ~default:0 body in
+          [("Content-length", Int.to_string length)]
+        end
+      | _ -> [] in
+    let headers = headers @ [("Host", host_str)] in
+    let (amz_headers, hashed_payload) = Auth.make_amz_headers time ?body in
+    let headers = headers @ amz_headers in
+    let request = Request.make ~meth
+        ~headers:(Header.of_list headers)
+        uri in
+    let auth_header = Auth.auth_request ~now:time
+        ~hashed_payload ~region:conf.region ~service:`S3
+        ~aws_access_key:conf.aws_access_key
+        ~aws_secret_key:conf.aws_secret_key request
+    in
+    let headers' = headers @ auth_header in
+    Cohttp.Request.(request.headers <- (Header.of_list headers'));
+    match meth with
+    | `PUT -> Client.request
+                ~body:(Option.value ~default:"" body |> Body.of_string)
+                request
+    | `GET -> Client.request request
+    | _ -> failwith "not possible right now"
+end
+
+type s3path = {bucket : string; objekt : string}
+
+type cmd =
+    S3toLocal of s3path * string
+  | LocaltoS3 of string * s3path
+
+let determine_s3_parts s =
+  (* Takes: string of the form s3://<bucket>/<object> *)
+  let s = String.drop_prefix s 5 in
+  let parts = String.split ~on:'/' s in
+  match parts with
+  | bucket::rst -> {bucket=bucket; objekt=(String.concat ~sep:"/" rst)}
+  | _ -> failwith "error format must be 's3://<bucket>/<object>'"
+
+let determine_paths src dst =
+  let is_s3 s = String.is_prefix ~prefix:"s3://" s in
+  match is_s3 src, is_s3 dst with
+  | (true, false) -> S3toLocal (determine_s3_parts src, dst)
+  | (false, true) -> LocaltoS3 (src, determine_s3_parts dst)
+  | (false, false) -> failwith "Use cp(1) :)"
+  | (true, true) -> failwith "Does not support copying from s3 to s3"
+
+let run region_str aws_access_key aws_secret_key src dst () =
+  (* nb client does not support redirects or preflight 100 *)
+  let open S3 in
+  let region = region_of_string region_str in
+  let conf = {region; aws_access_key; aws_secret_key} in
+  match determine_paths src dst with
+  | S3toLocal (src, dst) ->
+    begin
+      make_request conf ~meth:Code.(`GET) ~bucket:src.bucket ~objekt:src.objekt
+      >>= fun (resp, body) ->
```

Comment:
There's the Body.to_string below that. So it seems that one of them can be >>| but the other needs to be >>=. Regardless, it cleans up the code. Thanks!